### DISCUSSION TITLE
Wiki page to collect ideas for future development direction?

### DISCUSSION THREAD

**Discussion by simonpickering** [Navid200/xDrip#94] [Ideas]
I guess this discussion forum is the place to actually suggest and discuss those ideas, but is there a wiki page available where the outcome of those discussions could be prettified and collated? - perhaps something for the future.

I'll have a dig through the comments I've made on issues (which is not where it should have been done, I realise, but at the time there wasn't really a dedicated chat area like there is here - so thanks @Navid200 :)) and I'll add a new discussion to this section with some of my thoughts as to where I'd like to see things go.

**Comment by Navid200** [comment]
Yes, I hope that we can use these discussions for all sorts of conversations about xDrip.

With respect to development, one concern of mine that sometimes does not get attention is maintenance.  
There are times a change causes to break the code, a catastrophic failure.  Many people complain and we have to figure out a way to fix it.  This is a good scenario in my opinion!
The bad scenario is when a change causes a problem that is not catastrophic.  It may result in issues that no one notices at the time.
By the time, a few individuals complain about the problem, the person who made the change may not be an active developer any longer.  
A negative impact that is not catastrophic could stay in the code undetected.

To avoid, or at least minimize, this, we have to limit changes to only changes that are absolutely necessary.  That I believe is the reason many of the open pull requests are not merged.

In my opinion, please don't spend a lot of time developing something before spending some time talking about it with the developers.  That's one of the reasons I hope everyone would start using these discussions more and more.

Thanks for using the discussions.

**Comment by simonpickering** [comment]
It would be great to get input from the developers who know the code and the issues around implementing features, and I realise that there is a desire/need, to ensure that XDrip+ remains stable (the bugs and transient developers comment is pertinent especially for large changes).

From my point of view, there are features that I would really rather like and that I've seen dismissed out of hand, and I'm not sure if that's because they've been tried before and didn't work, don't fit with the current direction of the project (whatever that may be), don't fit with what a developer thinks. 

I hope, as you say, these discussions can be used to work out which of these is the issue - and ensuring that people can try out their ideas but without impacting others who are happy with the status quo.

Is there a willingness to host multiple branches to test whether certain features would work/be useful or not, or should this be done via our own repos? (And I've no idea what one does about naming at that point if functionality diverges!)

**Comment by Navid200** [comment]
This is a slide from someone else's presentation, not mine, that was shared with me.
![241717742_1072805803454606_1589354950826589978_n](https://user-images.githubusercontent.com/51497406/139948068-7c08c434-d85e-43ec-a461-5059c65818c5.png)
I like it because I can see I'm not the only one who has observed the problem.  
The concept of horse work versus donkey work is what I see here a lot.
There are developers who like to add new features.  But, not many developers like to spend time on fixing these bugs:
https://github.com/Navid200/xDrip/wiki/Bugs
We need to fix them.  But, not many people are aware that we are fixing these.  
There is work going on continuously on translations in the background so that people can use xDrip in their language.  Not many people are aware of this work.  These are all the donkey work, which must be done.  But, they are not glorified.  Nevertheless, they are very important.

The more features are added by different people with different styles, the more bugs we will have in the future.
This is partly the reason I have stated an opposing opinion to some new feature requests.  My thinking is that we should add a new feature only if absolutely necessary.

I wish more people would come forward with ideas for fixing the bugs.

But, I am only one of the volunteers.  I don't make the last call on anything here.  All of this is just my personal view.

**Comment by simonpickering** [comment]
It's always been this way, and writing documentation comes even lower down the list typically.

I understand what you're saying, it's interesting to read that people have come and added code (to Nightscout) and then gone. One sort of assumes that unlike other projects where interest or need may wane, you're tied to it for life if you're diabetic and there are no other viable options. 

My counter would be that if people can add features they will become invested in the project and then motivated and able to fix bugs that annoy them/they have the understanding to address. The code is large and dare I say rather unstructured, so it takes some significant effort to work out what is going on with it (i.e. how to integrate your pet feature is a good motivator) before you know enough to know how to address other bugs. 

If pet features could be implemented in branches, for example, it would keep the main code clean (unless a merge is seen as being a good thing to expose the masses to some new functionality), and allow people to test their ideas and gain understanding of the code so they can and want to do other useful bug hunting.

This approach probably already happens (I see lots of branches), but I do feel the blanket rejection of ideas doesn't help - tell people to clone the repo and create a branch, encourage them to do this and test their code and say that it might be pulled back into mainline/master if it's demonstrated to be useful for the majority of users (and if not it's still there for the developer to use as they desire).

I'm not even a volunteer, having not contributed any code yet, so likewise it's just my opinion.

**Comment by Navid200** [comment]
This is open source.  That means anyone can create their own version.  There are other repositories.
What is unique about this repository is that it's not meant to make only one person happy.  It's rather for everyone who has diabetes worldwide.  There are many users who are not programmers and cannot create their own version.  But, it's more than that.  There are users who are not even tech savvy.   They use this repository and this repository keeps them in mind.

Those requirements are very different than the requirements for someone who likes to add a feature to xDrip to make it fit their own needs.  Some of the requirements are contradictory.

If I was in charge, I would tell any developer coming forward with a new feature development to fix a bug first before we would consider their request.  
Or, I would ask the developer to contribute to customer service helping users bring up their systems and help 50 users before we would consider the request.
Or, I would ask them to put up a wiki page about something missing in the current documentation before we would consider their request.  It's only fair.  
But, I'm not.

We already have a list of feature requests:  https://github.com/Navid200/xDrip/wiki/Feature-Requests
We even have a load of already developed features and bug fixes:  https://github.com/NightscoutFoundation/xDrip/pulls
It's not like we have a shortage of ideas.  

Have you seen this bug?
https://github.com/NightscoutFoundation/xDrip/issues/1266
It is unacceptable to have a bug like this.  A user posted today on facebook saying she was worried if this would happen when her kid was at school.  
It makes me sad.  How can we have a bug like this?  It was reported 1.5 years ago.  We still cannot fix it.  The reason we cannot fix it is that as far as I know no one knows for sure exactly what causes it.  Why would that be the case if xDrip wasn't as complicated as it is today?  
Adding more features will only make it more complicated.
And then, there are developers who want to add a bolus calculator instead of spending time on fixing that bug or others like it.  There are people who ask why xDrip cannot upload pictures of their food!
There are users who want xDrip to detect their Bluetooth headphone and automatically enable them.  
...
